src/syntax.rs

```diff
@@ -0,0 +1,105 @@
+pub enum DiscreteType {
+    Int,
+    Bool,
+    Float,
+}
+
+pub enum Type {
+    Vector(usize, DiscreteType),
+    Function,
+    Shader,
+}
+
+pub enum ShaderType {
+    Vector,
+    Fragment,
+    Tesselation,
+    // And more!
+}
+
+pub enum BOp {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    Mod,
+    And,
+    Or,
+    Eq,
+    Neq,
+    Gt,
+    Gte,
+    Lt,
+    Lte,
+    BitAnd,
+    BitOr,
+    BitXor,
+}
+
+pub struct Block {
+    pub body: Vec<Box<Expr>>
+}
+pub struct Parameter {
+    name: String,
+    datatype: Type
+}
+
+pub enum AccessType {
+    Name(String),
+    Idx(usize),
+}
+
+pub enum Expr {
+    I(usize),
+    B(bool),
+    F(f32),
+    Vect {
+        is_matrix: bool,
+        datatype: DiscreteType,
+        value: Block,
+    },
+    Def {
+        mutable: bool,
+        name: String,
+        value: Box<Expr>,
+    },
+    Ref(String),
+    App {
+        fname: String,
+        arguments: Block,
+    },
+    BinOp {
+        operator: BOp,
+        e1: Box<Expr>,
+        e2: Box<Expr>,
+    },
+    Compose(String, String),
```

Comment:
This allows composition of named functions, but not for unnamed functions (ex. lambdas). If we allow unnamed functions, we probably will need to change this.

Reply:
Very true! In the old AST, `Compose` was a `BinOp`. I think I agree with you that this should be the case.

Reply:
This also brings up the question, how are we representing composition of shaders in the AST?

Reply:
Assuming functions are still first class values, it would just be a binary operation on functions as values. With the way this syntax is written, we could drop the 'name' property from `Func`, and be able to create unnamed functions that could be bound to a name via `Def` (or `MutDef`). Then I think that should work okay 🤔